watertap/property_models/ion_DSPMDE_prop_pack.py

```diff
@@ -640,21 +791,81 @@ def define_state_vars(self):
+        if adjust_by_ion is not None:
+            if adjust_by_ion in self.params.solute_set:
+                self.charge_balance = Constraint(expr=sum(self.charge_comp[j] * self.conc_mol_phase_comp['Liq', j]
+                                                 for j in self.params.solute_set) == 0)
+            else:
+                raise ValueError('adjust_by_ion must be set to the name of an ion in the list of solutes.')
+        if defined_state:
             for j in self.params.solute_set:
-                if not self.flow_mol_phase_comp['Liq', j].is_fixed():
+                if not self.flow_mol_phase_comp['Liq', j].is_fixed() and adjust_by_ion != j:
                     raise AssertionError(
                         f"{self.flow_mol_phase_comp['Liq', j]} was not fixed. Fix flow_mol_phase_comp for each solute"
                         f" to check that electroneutrality is satisfied.")
-        val = value(sum(self.charge_comp[j] * self.flow_mol_phase_comp['Liq', j]
-                    for j in self.params.solute_set))
+                if adjust_by_ion == j and self.flow_mol_phase_comp['Liq', j].is_fixed():
+                    self.flow_mol_phase_comp['Liq', j].unfix()
+        else:
+            for j in self.params.solute_set:
+                if self.flow_mol_phase_comp['Liq', j].is_fixed():
+                    raise AssertionError(
+                        f"{self.flow_mol_phase_comp['Liq', j]} was fixed. Either set defined_state=True or unfix "
+                        f"flow_mol_phase_comp for each solute to check that electroneutrality is satisfied.")
+
+        # touch this var since it is required for this method
+        self.conc_mol_phase_comp
+
+        if solve:
+            solve = get_solver()
+            results = solve.solve(self)
+            if check_optimal_termination(results):
+                val = value(sum(self.charge_comp[j] * self.conc_mol_phase_comp['Liq', j]
+                            for j in self.params.solute_set))
+            else:
+                if adjust_by_ion is not None:
+                    del self.charge_balance
+                raise ValueError('The stateblock failed to solve while computing concentrations to check the charge balance.')
+        else:
+            val = value(sum(self.charge_comp[j] * self.conc_mol_phase_comp['Liq', j]
+                            for j in self.params.solute_set))
         if abs(val) <= tol:
+            if adjust_by_ion is not None:
+                del self.charge_balance
+                ion_adjusted = self.flow_mol_phase_comp['Liq', adjust_by_ion].value
+                if defined_state:
+                    self.flow_mol_phase_comp['Liq', adjust_by_ion].fix(ion_adjusted)
+                    # touch on-demand property desired
+                    if get_property is not None:
+                        if isinstance(get_property, str):
+                            getattr(self, get_property)
+                        elif isinstance(get_property, list):
```

Comment:
```suggestion
                        elif isinstance(get_property, (list, tuple)):
```
We might make this slightly more general. But it's also fine as-is.


I guess we could also use a different ask-for-forgiveness pattern:
```python
try:
    getattr(self, get_property)
except AttributeError: #assume a list or other iterator of properties
    for i in get_property:
        getattr(self,i)
except:
    raise TypeError("get_property must be a string or an iterable of strings.")
```